frontend/class_node.py

```diff
@@ -42,8 +42,8 @@ def all_parents(self):
         Returns all transitive parent nodes.
         """
         result = [self]
-        if self.parent_node:
```

Comment:
The result could return duplicates, right? I don't think that would break anything, but could we prevent that (preferably without introducing non-determinism)?